apps/settings/js/modules/bluetooth/bluetooth_connection_manager.js

```diff
@@ -0,0 +1,408 @@
+        });
+      });
+    },
+
+    /**
+     * The method will disconnect the input device with the device's adapter.
+     *
+     * @access public
+     * @memberOf BluetoothConnectionManager
+     * @param {Object BluetoothDevice} device
+     * @returns {Promise}
+     */
+    disconnect: function btcm_disconnect(device) {
+      return new Promise((resolve, reject) => {
+        if (!this._defaultAdapter) {
+          reject('default adapter is not existed!!');
+        }
+
+        var req = this._defaultAdapter.disconnect(device);
+        req.onsuccess = () => {
+          this.connectedAddress = null;
+          Debug('disconnect(): connectedAddress = null');
+          Debug('disconnect(): onsuccess(): resolve');
+          resolve();
+        };
+
+        req.onerror = () => {
+          Debug('disconnect(): onerror(): reject with disconnect failed');
+          reject('disconnect failed');
+        };
+      });
+    },
+
+    /**
+     * The method will disconnect the input device with the device's adapter.
+     *
+     * @access public
+     * @memberOf BluetoothConnectionManager
+     * @param {String} address
+     * @returns {Promise} resolve: profiles
+     * @returns {Promise} reject: reason
+     */
+    fetchDeviceConnectionProfiles: 
+    function btcm_fetchDeviceConnectionProfiles(address) {
+      return new Promise((resolve, reject) => {
+        if (!this._defaultAdapter) {
+          reject('default adapter is not existed!!');
+        }
+
+        Debug('fetchDeviceConnectionProfiles(): address = ' + address);
+        // Get connected devices from platform side.
+        this._getConnectedDevices().then((connectedDeviceItemsMap) => {
+          var profiles = connectedDeviceItemsMap[address].connectedProfiles;
+          Debug('fetchDeviceConnectionProfiles(): profiles = ' + 
+                JSON.stringify(profiles));
+          Debug('fetchDeviceConnectionProfiles(): resolve with profiles');
+          resolve(profiles);
+        }, (reason) => {
+          Debug('fetchDeviceConnectionProfiles(): ' + 
+                'reject with reason: ' + reason);
+          reject(reason);
+        });
+      });
+    },
+
+    /**
+     * The method will get all connected devices profiles 
+     * which we are interested in. Profile: HFP, A2DP.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @returns {Promise} resolve: profiles
+     * @returns {Promise} reject: reason
+     */
+    _getConnectedDevices: function btcm__getConnectedDevices() {
+      return new Promise((resolve, reject) => {
+        // get connected device via Profiles.HFP
+        this._getConnectedDevicesByProfile(
+          this.Profiles.HFP).then((connectedDevices) => {
+          Debug('_getConnectedDevices(): connectedDevices = ' + 
+                JSON.stringify(connectedDevices));
+          // Update each connected devices in map.
+          this._updateDeviceItemsMap(this.Profiles.HFP, connectedDevices);
```

Comment:
This is confusing to me. It seems you always get the connected devices from gecko and construct the map everytime. If you would like to use `_connectedDeviceItemsMap` as a cache, you should return the cache directly and update the cache when there are change events from gecko notifying the connection change (although I'm not sure if such event exists).